src/Models/DeviceDetails.php

```diff
@@ -0,0 +1,79 @@
+<?php
+namespace Ably\Models;
+
+function get($arr, $key) {
+    foreach(explode('.', $key) as $k) {
+        if (is_null($arr) || ! isset($arr[$k])) {
+          return NULL;
+        }
+        $arr = $arr[$k];
+    }
+
+    return $arr;
+}
+
+class DeviceDetails extends BaseOptions {
+
+    const DevicePushTransportType = ['fcm', 'gcm', 'apns', 'web'];
```

Comment:
We've tended not to include this kind of validation check in the library, because then if we do add support for new types it's not possible to access that functionality without having a new library. I think it's best that the backend performs this validation.